### ISSUE TITLE
Mixnet key storage and management

### ISSUE THREAD

**Issue by zdave** [paritytech/substrate#12595]
Each mixnet node will need a key pair for Diffie-Hellman key exchange with message senders. The public keys of all mixnet nodes will need to be known by all message senders. Ideally the mixnet nodes would rotate their keys automatically every few hours to provide forward security.

This issue is to discuss how these keys should be stored and managed.

For the initial mixnet release, the current plan is to use a session key, which would be managed and distributed the same way as the other session keys. In particular, it would need to be manually rotated.

Ultimately the plan is to automatically rotate the keys. A node might do this by using a session key to sign the new Diffie-Hellman public key and putting this on-chain with an unsigned transaction.

**Comment by bkchr** [comment]
> The public keys of all mixnet nodes will need to be known by all message senders.

Will only validators be mixnet nodes? Or can everybody be a mixnet node? And why do we need to know all public keys? We are doing a key exchange.

> For the initial mixnet release, the current plan is to use a session key, which would be managed and distributed the same way as the other session keys. In particular, it would need to be manually rotated.

If only validators will be mixnet nodes, this can work. However, when we want to have all nodes to be mixnet nodes, we can not really use session keys. The question would also be, do we need to proof to someone that for example I'm owning a particular public key to some on chain logic?

I think in general this issue needs to have some more information on how the entire mixnet system should work. :)

**Comment by burdges** [comment]
Agreed.  We need some mechanism to choose who is a mix node, but not necessarily our validator selection criteria.  

In fact, we'd make the mix net software separate from substrate in an ideal world, but doing so would create a lot of work for people to set it up.  It's kinda tricky..

**Comment by xlc** [comment]
what is mixnet?

**Comment by bkchr** [comment]
https://en.wikipedia.org/wiki/Mix_network

**Comment by burdges** [comment]
Mixnets provide anonymity using onion encryption like Tor, but each packet is routed separately.

Tor should be more secure if you've a stream of traffic, like TCP, HTTP, etc. because each packet taking different routes means your stream quickly hits a route where all nodes are malicious.

All known theoretical models of anonymity need both latency and cover traffic, so folks think Tor does not benefit from adding cover traffic.

Mixnets should be more secure if you've sparser messages and replies, like email, messangers, git, submitting tx, etc.  Mixnets intentionally add latency, which messes up TCP, but adding latency makes means adding cover traffic too gives mixnets better anonymity.

**Comment by zdave** [comment]
> Will only validators be mixnet nodes? Or can everybody be a mixnet node? And why do we need to know all public keys? We are doing a key exchange.

For the initial release (AIUI) we're just planning to support anonymous submission of transactions on the Polkadot relay chain. The simplest option is for the mixnet node set to be the validator set, so that's what we were planning to go with, for now at least.

> And why do we need to know all public keys? We are doing a key exchange.

To send a message through the mixnet you need to know the public keys of all the nodes on the path, as you do a key exchange with each one. Maybe there is some mechanism by which you could query just these keys without revealing anything about the message path, but the simplest option is to just have every sending node know all the mixnet node keys.

> The question would also be, do we need to proof to someone that for example I'm owning a particular public key to some on chain logic?

AIUI the anonymity of the network is easier to reason about if everyone is using the same set of mixnet nodes. This is easy to achieve if the mixnet node set is determined by on-chain logic. I think this is the main reason for this stuff to go on-chain; I don't think for example that the node set needs to be known by any other on-chain logic.

If you're sending a message you need to know with some certainty that the public keys for the nodes along the path really do belong to the nodes you think they do. If for example the first node knows all the corresponding private keys then they could just completely unwrap your message. So you do need to prove the public key <-> mixnet node relation to sending nodes in some way.

I've possibly misunderstood this question...

**Comment by bkchr** [comment]
> I've possibly misunderstood this question...

No, your explanations are already on the correct path :P However, personally I still miss an overall explanation on what you need and how it should work to help here.

**Comment by burdges** [comment]
At least some of libp2p's abstractions sound pretty dangerous for anonymity stuff, so maybe we should consider basing this upon smaldot's networking.  I'd actually suggest exploring whether it made sense to add to add to arti, except tor won't be done with relay support in arti for another year.

**Comment by zdave** [comment]
> At least some of libp2p's abstractions sound risky for anonymity stuff

Could you elaborate on this?

**Comment by burdges** [comment]
I'm firstly worried about learning peer information from the DHT, which afaik is how libp2p works, and how it comes up here.

**Comment by zdave** [comment]
> I'm firstly worried about learning peer information from the DHT, which afaik is how libp2p works, and how it comes up here.

Hmm ok. For forwarding messages the plan is just to piggy-back on the authority discovery component, which AIUI already looks up and connects to all the nodes in the current (and next) validator set. For nodes sending messages I think they would need to look up the peer information for the first hop, but not for subsequent hops.

**Comment by zdave** [comment]
> I still miss an overall explanation on what you need and how it should work to help here.

So the plan is to use the [Sphinx message format](https://cypherpunks.ca/~iang/pubs/Sphinx_Oakland09.pdf). The sender of a message determines the path of the message through the mixnet up front. It builds the message by repeatedly encrypting and extending, starting with the final hop and working backwards, so that each hop can only see the information relevant to it. The key used to encrypt/decrypt at each hop is derived from the keys of the sender and the hop node by Diffie-Hellman key exchange.

The sender can generate a new key pair for every message and essentially includes the public key in the message. The key pair for each hop node needs to be generated in advance though, and the public key needs to be distributed, as it needs to be known by the message sender when building the message.

Ideally these keys would be rotated fairly regularly, and the old private keys destroyed, as if private keys get leaked they can be used to deanonymise messages.
